### ISSUE TITLE
React Hooks for styles and theming 

### ISSUE THREAD

**Issue by hraz-msft** [microsoft/fluentui#14622]
#### Describe the feature that you would like added

Please add React Hooks for styling like in React-JSS or Material-UI libraries, that will allow to access the theme object directly and release the dependencies of 3rd party libraries.

#### What component or utility would this be added to

Not relevant...

#### Have you discussed this feature with our team, and if so, who

No

#### Additional context/screenshots

Something like:

```
const useStyles = fluentUseStyles((theme: ITheme) => ({
    root: {
        color: theme.palette.themeLighterAlt,
        fontSize: theme.fonts.superLarge,
    }
});

const MyButton: React.FC = () => {
    const classes = useStyles();

    return <DefaultButton classNames={classes.root} />
}
```

**Comment by paulgildea** [comment]
@hraz-msft Thanks for the feature request.

@dzearing FYI who's been driving styling all up for the component library.

At this time we are examining our styling API. I don't have any specific timelines to share at this time, but we'll take this into consideration as we make progress. 

Feel free to reach out if you want to play around the API once we have something consumable for feedback.

Marking this as "Won't fix" at this time because are we restructuring the API.

**Comment by dzearing** [comment]
We are going to provide an interim solution to help here. I think it should help unblock your use cases.

You can preview the implementation of `useStyles` in the codesandbox here:

https://codesandbox.io/s/usestyles-to-build-styling-using-the-fluent-ui-theme-wt0ue?file=/src/App.tsx

However, as @paulgildea mentioned, we're implementing a more performant approach. Rendering styling like this has some performance consequences that aren't terrible but can certainly be abused when done in bulk. We want to do better.

Some of the pitfalls we've seen:

* Every render must compute the styling
* You can be clever memoizing the result, but the memo keys must be fast to calculate and not constantly mutate.
* Memoizing should be done on a more global level rather than instance level, otherwise you're not saving time.

Which means we need to have something that scales a bit better than per instance evaluation of styling.

**Comment by paulgildea** [comment]
Love it! Thanks @dzearing glad to hear we could provide some value for v7 here!

**Comment by hraz-msft** [comment]
Thank you! Looks awesome! I can't wait to see the official result!

My suggestion was to calculate the style once outside of the component, but then the theme cannot be changed without full refresh (for example if the user want to move from light theme to dark theme and vice versa).

If you are using the `useStyles` inside of the component rendering logic, you can use the `useMemo` hook to memorize the "expensive" calculation of the style, and use the theme object as a dependency. This is a quick way to improve the performance a little bit 😄

**Comment by dzearing** [comment]
Totally thought about `useMemo`. The problem is that it will memo on a per component instance, so each of your buttons will individually run the function, when you want to have more of a global cache which only calculates classes 1 time for a unique theme object.

**Comment by hraz-msft** [comment]
Hi @dzearing, I saw that the PR was merged! 😄 

When it will be released as npm package?

**Comment by dzearing** [comment]
Overnight at 3am.

**Comment by dzearing** [comment]
@hraz-msft 

Example usage here:

https://codesandbox.io/s/fluent-ui-themeprovider-and-makestyles-example-d84my

This was released in the @fluentui/react-theme-provider package. Should work with existing theme via loadTheme or with Customizer provided themes. Please let us know if you run into any issues (feel free to reach out on Teams.)

**Comment by hraz-msft** [comment]
Thanks @dzearing, I will integrate it to my project tomorrow morning! Thank you for the update and the rapid development!
